types/prefs/map.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"maps"
+	"net/netip"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"golang.org/x/exp/constraints"
+	"tailscale.com/types/opt"
+	"tailscale.com/types/ptr"
+	"tailscale.com/types/views"
+)
+
+// MapKeyType is a constraint allowing types that can be used as [Map] and [StructMap] keys.
+// To satisfy this requirement, a type must be comparable and must encode as a JSON string.
+// See [jsonv2.Marshal] for more details.
+type MapKeyType interface {
+	~string | constraints.Integer | netip.Addr | netip.Prefix | netip.AddrPort
```

Comment:
maybe this?
```suggestion
	cmp.Ordered | netip.Addr | netip.Prefix | netip.AddrPort
```

Reply:
We have [special handling](https://github.com/go-json-experiment/json/blob/ebd3a8989ca1eadb7a68e02a93448ecbbab5900c/arshal.go#L103-L107) for numeric keys in the `jsonv2` package:
```
//   - A Go map is encoded as a JSON object, where each Go map key and value
//     is recursively encoded as a name and value pair in the JSON object.
//     The Go map key must encode as a JSON string, otherwise this results
//     in a [SemanticError]. When encoding keys, [StringifyNumbers]
//     is automatically applied so that numeric keys encode as JSON strings.
```
Even though the `cmp.Ordered` constraint allows only strings and numbers today today, it is semantically different, and the doc says:

> Ordered is a constraint that permits any ordered type: any type that supports the operators < <= >= >.
> If future releases of Go add new ordered types, **this constraint will be modified to include them.**

I'd also prefer not to allow floating-point numbers as the key, hence `constraints.Integer`.